File: scripts/audit_installed_palette.py
"""Audit installed Neovim plugins against Dawn's palette without updating plugins."""
import argparse
import os
from pathlib import Path
import subprocess
import tempfile

ROOT = Path(__file__).resolve().parents[1]


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--nvim', default='nvim')
    parser.add_argument('--init', type=Path, default=Path.home()/'.config/nvim/init.lua')
    parser.add_argument('--lazy-root', type=Path, default=Path.home()/'.local/share/nvim/lazy/lazy.nvim')
    parser.add_argument('--output', type=Path, required=True)
    args = parser.parse_args()
    if not args.init.is_file() or not args.lazy_root.is_dir():
        parser.error('Installed Neovim config and lazy.nvim are required; coverage is unavailable.')
    output = args.output.resolve()
    output.mkdir(parents=True, exist_ok=True)
    with tempfile.TemporaryDirectory(prefix='ithilien-plugin-audit-') as tmp:
        env = dict(os.environ, ITHILIEN_ROOT=str(ROOT), ITHILIEN_NVIM_INIT=str(args.init.resolve()),
                   ITHILIEN_LAZY_ROOT=str(args.lazy_root.resolve()), ITHILIEN_AUDIT_OUTPUT=str(output),
                   XDG_STATE_HOME=tmp+'/state', XDG_CACHE_HOME=tmp+'/cache')
        with (output/'nvim.log').open('w') as log:
            result = subprocess.run([args.nvim, '--headless', '-u', str(ROOT/'scripts/audit_installed_palette.lua'),
                                     '-i', 'NONE', '-n'], cwd=tmp, env=env, stdout=log, stderr=subprocess.STDOUT, timeout=120)
    print(f"{'PASS' if result.returncode == 0 else 'FAIL'}: installed highlight audit; {output/'result.json'}")
    return result.returncode


if __name__ == '__main__':
    raise SystemExit(main())

File: scripts/audit_installed_palette.lua
-- Run through audit_installed_palette.py; uses the actual user config without installing plugins.
vim.opt.rtp:prepend(assert(vim.env.ITHILIEN_LAZY_ROOT, 'Set ITHILIEN_LAZY_ROOT'))
local lazy=require('lazy');local setup=lazy.setup
lazy.setup=function(o)
 o.install={missing=false};o.checker={enabled=false};o.change_detection={enabled=false};o.readme={enabled=false}
 o.lockfile=vim.env.ITHILIEN_AUDIT_OUTPUT..'/lock.json';o.state=vim.env.ITHILIEN_AUDIT_OUTPUT..'/state.json';table.insert(o.spec,{ 'achandran/ithilien', dir=assert(vim.env.ITHILIEN_ROOT, 'Set ITHILIEN_ROOT') });return setup(o)
end
dofile(assert(vim.env.ITHILIEN_NVIM_INIT, 'Set ITHILIEN_NVIM_INIT'))
vim.schedule(function()
 local r={plugins={},snapshots={}}
 local s=vim.json.decode(table.concat(vim.fn.readfile(vim.env.ITHILIEN_ROOT..'/scripts/palette/ithilien.json'),'\n')).variants['ithilien-dawn']
 r.palette=s.colors;r.theme=vim.g.colors_name
 local allowed={};for _,v in pairs(s.colors) do allowed[tonumber(v:sub(2),16)]=true end
 local function snap(label)
  local bad={};local count=0
  for n,h in pairs(vim.api.nvim_get_hl(0,{})) do
   count=count+1
   for _,a in ipairs({'fg','bg','sp'}) do if h[a] and not allowed[h[a]] then bad[#bad+1]={group=n,attribute=a,color=string.format('#%06X',h[a])} end end
  end
  r.snapshots[#r.snapshots+1]={label=label,groups=count,off_palette=bad}
 end
 snap('startup')
 local skip={catppuccin=true,['tokyonight.nvim']=true,['dark-notify']=true}
 for n,p in pairs(require('lazy.core.config').plugins) do
  local status
  if skip[n] then status='excluded alternate theme or appearance watcher'
  elseif vim.fn.isdirectory(p.dir)==0 then status='missing'
  else local ok,e=pcall(lazy.load,{plugins={n}});status=ok and (p._.loaded and 'loaded' or 'not loaded') or tostring(e) end
  r.plugins[#r.plugins+1]={name=n,status=status}
 end
 vim.wait(100);snap('plugins loaded');vim.cmd('colorscheme ithilien-dawn');vim.wait(100);snap('theme reapplied')
 vim.fn.writefile({vim.json.encode(r)},vim.env.ITHILIEN_AUDIT_OUTPUT..'/result.json');local failed=false
 for _,stage in ipairs(r.snapshots) do if #stage.off_palette>0 then failed=true end end
 for _,plugin in ipairs(r.plugins) do if plugin.status~='loaded' and not plugin.status:match('^excluded') then failed=true end end
 vim.cmd(failed and 'cquit 1' or 'qa!')
end)
